File: app/models/game/building/surface.rb
module Game
  module Building
    # One flat face of a building, as a grid of cells: a storey of one exterior wall, an
    # interior partition, a floor deck, a roof plane, a gable end.
    #
    # The grid is the whole trick. A building is not stored as pieces and is never sent as
    # pieces -- it is sent as a handful of surfaces, and the cells are implied. Twenty or
    # so surfaces describe what would otherwise be two hundred and fifty pieces, and the
    # client expands them with arithmetic that cannot drift because there is nothing in it
    # to get wrong.
    #
    # THE RULE THAT MAKES THAT SAFE: piece index space is never culled. Every row and
    # column has an index whether or not anything is drawn there. A doorway is a real
    # index holding `void`; a gable's clipped corners are real indices holding `void`. If
    # culling changed the numbering, Ruby and JavaScript would both have to cull
    # identically forever, and the first divergence would silently renumber every piece in
    # the building -- so damage recorded against one wall would come back applied to
    # another.
    class Surface
      KINDS = %i[wall partition floor roof gable rubble].freeze

      attr_reader :kind, :storey, :material, :origin, :u, :v, :normal,
                  :width, :height, :cols, :rows, :thickness, :patches, :piece_offset, :seed

      def initialize(kind:, storey:, material:, origin:, u:, v:, width:, height:,
                     cols:, rows:, thickness:, patches: [], piece_offset: 0, seed: 0)
        @kind = kind
        @storey = storey
        @material = material
        @origin = origin
        @u = u
        @v = v
        @normal = u.cross(v).normalised
        @width = width.to_f
        @height = height.to_f
        @cols = cols
        @rows = rows
        @thickness = thickness.to_f
        @patches = patches
        @piece_offset = piece_offset
        @seed = seed
      end

      def piece_count = cols * rows

      def piece_index(row, col)
        piece_offset + row * cols + col
      end

      # Inverts piece_index. Used by the server to find which surface a reported index
      # belongs to, and by tests to prove the two agree.
      def local_index(piece_index)
        piece_index - piece_offset
      end

      def covers?(piece_index)
        local = local_index(piece_index)
        local >= 0 && local < piece_count
      end

      def cell_width = width / cols
      def cell_height = height / rows
      def cell_area = cell_width * cell_height

      # How much of this surface actually holds something up: cell area weighted by the
      # material in each cell, so a window and an empty doorway weigh nothing. With a
      # block, only the cells it accepts are counted -- which is how the collapse rule
      # asks what is left standing without a second traversal that could disagree.
      def structural_area
        rows.times.sum do |row|
          cols.times.sum do |col|
            next 0.0 if block_given? && !yield(piece_index(row, col))

            material_at(row, col).structural_weight * cell_area
          end
        end
      end

      # The material at one cell: the last patch covering it wins, so a lintel laid over a
      # window opening reads as the lintel.
      def material_at(row, col)
        patch = patches.reverse.find { |p| p.covers?(row, col) }
        patch ? Materials.fetch(patch.material) : material
      end

      # Every material that appears anywhere on this surface.
      def materials
        ([ material ] + patches.map { |patch| Materials.fetch(patch.material) }).uniq(&:name)
      end

      def per_material
        materials.to_h { |m| [ m.name.to_s, yield(m) ] }
      end

      def with_offset(offset)
        self.class.new(
          kind: kind, storey: storey, material: material, origin: origin, u: u, v: v,
          width: width, height: height, cols: cols, rows: rows, thickness: thickness,
          patches: patches, piece_offset: offset, seed: seed
        )
      end

      # Which cells break together. Worked out after the offset is known, because the
      # tiling is seeded from it -- two surfaces with identical grids should not come out
      # with identical blocks.
      def blocks
        @blocks ||= Blocks.tile(self, seed: seed)
      end

      def to_spec
        {
          kind: kind.to_s,
          storey: storey,
          mat: material.name.to_s,
          o: origin.to_a,
          u: u.to_a,
          v: v.to_a,
          n: normal.to_a,
          w: width,
          h: height,
          cols: cols,
          rows: rows,
          t: thickness,
          off: piece_offset,
          # The client lays rubble out from this. Two players seeing heaps in the same
          # place depends entirely on both deriving them from the same seed, so a surface
          # that did not carry its own seed would have every building's wreckage laid out
          # identically -- which passes for one building and fails for a street.
          seed: seed,
          # Per cell, per material -- not per surface. A window is a glass cell in a brick
          # wall, and giving it the wall's health would make it as hard to break as the
          # wall, which is the opposite of the point. Keyed by name and covering the
          # surface's own material plus every material any patch introduces.
          hp: per_material { |m| m.health_for(cell_area, thickness) },
          kg: per_material { |m| m.mass_for(cell_area, thickness) },
          str: per_material(&:structural_weight),
          # One block id per cell, row-major, or absent where the plain grid is right.
          blocks: blocks,
          patches: patches.map(&:to_spec)
        }
      end

      # A rectangle of cells whose material differs from the surface's own: a window, a
      # door, a lintel, the clipped corner of a gable. Inclusive at both ends.
      class Patch
        attr_reader :col0, :row0, :col1, :row1, :material

        def initialize(col0:, row0:, col1:, row1:, material:)
          @col0 = col0
          @row0 = row0
          @col1 = col1
          @row1 = row1
          @material = material.to_sym
        end

        def covers?(row, col)
          row.between?(row0, row1) && col.between?(col0, col1)
        end

        def to_spec
          [ col0, row0, col1, row1, material.to_s ]
        end
      end
    end
  end
end

File: app/javascript/game/world/surface.js
import * as THREE from "three"

// Turning a surface back into the cells it describes.
//
// This is the other half of Game::Building::Surface, and the only new piece of logic that
// has to agree across the two languages. It is deliberately trivial and TOTAL: every row
// and column produces a cell, always. Nothing here decides that a cell should be skipped,
// because piece indices are counted from the grid -- if either side culled, the two would
// have to cull identically forever, and the first divergence would renumber every piece
// after it.
//
// A doorway is not a missing cell. It is a cell whose material is `void`, which the caller
// then declines to give a body or a mesh. Same index, same arithmetic, nothing to get
// wrong.

// Reused for every cell of every surface: expanding a building allocates nothing.
const u = new THREE.Vector3()
const v = new THREE.Vector3()
const n = new THREE.Vector3()
const position = new THREE.Vector3()
const scale = new THREE.Vector3()
const matrix = new THREE.Matrix4()
// Walked backwards, because the last patch covering a cell wins -- a lintel laid over a
// window opening has to read as the lintel.
export function materialAt(surface, row, col) {
  for (let i = surface.patches.length - 1; i >= 0; i -= 1) {
    const [ col0, row0, col1, row1, material ] = surface.patches[i]
    if (row >= row0 && row <= row1 && col >= col0 && col <= col1) return material
  }
  return surface.mat
}

export function pieceIndex(surface, row, col) {
  return surface.off + row * surface.cols + col
}

export function cellSize(surface) {
  return { width: surface.w / surface.cols, height: surface.h / surface.rows }
}

// The cell's transform, composed T * R * S and never T * S * R. The order matters: three
// derives an instance's normals by dividing out the squared column lengths of its matrix,
// which is the correct inverse-transpose only while the matrix carries no shear. Scaling
// before rotating introduces exactly that shear, and the lighting goes subtly wrong in a
// way that is very hard to attribute later.
// Cells sit flush: exactly coplanar, exactly abutting, all the same size.
//
// They did not, for a while. Each was nudged out of plane, rolled a couple of degrees and
// given its own shade, to stop a wall reading as graph paper -- and the result read as a
// patchwork of separate boxes stacked against each other, which is worse. A wall is one
// object that happens to be destructible; the irregularity belongs in how it comes APART,
// which is what the block tiling is for, not in how it looks standing up.
export function cellMatrix(surface, row, col, target, origin) {
  const width = surface.w / surface.cols
  const height = surface.h / surface.rows

  u.fromArray(surface.u)
  v.fromArray(surface.v)
  n.fromArray(surface.n)

  // Cells are centred on the surface plane, so a wall's thickness straddles the line its
  // origin describes rather than hanging off one face of it.
  position
    .fromArray(surface.o)
    .addScaledVector(u, (col + 0.5) * width)
    .addScaledVector(v, (row + 0.5) * height)

  if (origin) position.add(origin)

  target.makeBasis(u, v, n)
  target.scale(scale.set(width, height, surface.t))
  target.setPosition(position)
  return target
}

// The transform of a RECTANGLE of cells, for a slab that falls as one body. Composed
// exactly as cellMatrix composes one cell -- same basis, same centring on the surface
// plane, same T * R * S order and for the same shear reason -- so a slab starts life
// occupying precisely the space its cells did, down to the thickness straddling the plane.
//
// Its local axes are the surface's: x along u (columns), y along v (rows), z along the
// normal. Callers rely on that to place things inside the slab, which is how a landing
// slab throws one cell's worth of shards per cell it covered rather than one enormous
// fragment per slab.
export function chunkMatrix(surface, row, col, rows, cols, target, origin) {
  const width = surface.w / surface.cols
  const height = surface.h / surface.rows

  u.fromArray(surface.u)
  v.fromArray(surface.v)
  n.fromArray(surface.n)

  position
    .fromArray(surface.o)
    .addScaledVector(u, (col + cols / 2) * width)
    .addScaledVector(v, (row + rows / 2) * height)

  if (origin) position.add(origin)

  target.makeBasis(u, v, n)
  target.scale(scale.set(cols * width, rows * height, surface.t))
  target.setPosition(position)
  return target
}

// Every cell of a surface, in index order. `visit` receives the piece index, the material
// name, and a matrix it must not hold on to -- the same one is reused for every cell.
export function eachCell(surface, origin, visit) {
  for (let row = 0; row < surface.rows; row += 1) {
    for (let col = 0; col < surface.cols; col += 1) {
      visit(
        pieceIndex(surface, row, col),
        materialAt(surface, row, col),
        cellMatrix(surface, row, col, matrix, origin),
        surface
      )
    }
  }
}

export function eachBuildingCell(building, visit) {
  const origin = new THREE.Vector3().fromArray(building.o)
  for (const surface of building.surfaces) eachCell(surface, origin, visit)
}
